Lib/test/test_inspect.py

```diff
@@ -3674,6 +3674,43 @@ def test_signature_on_noncallable_mocks(self):
                 with self.assertRaises(TypeError):
                     inspect.signature(mock)
 
+    def test_signature_from_code(self):
+        def func1(
+            a, b: int,
+            /,
+            c, d: str = 'a',
+            *args: int,
+            e, f: bool = True,
+            **kwargs: str,
+        ) -> float:
+            ...
+
+        def func2(a: str, b: int = 0, /): ...
+        def func3(): ...
+        def func4(*a, **k): ...
+        def func5(*, kw=False): ...
+
+        known_sigs = {
+            func1: '(a, b, /, c, d, *args, e, f, **kwargs)',
+            func2: '(a, b, /)',
+            func3: '()',
+            func4: '(*a, **k)',
+            func5: '(*, kw)',
+        }
```

Comment:
Well, I worry that the reason this wasn't implemented in the original PEP-362 implementation is because the signatures produced for `func1`, `func2` and `func5` here are all incorrect, due to the fact that certain parameters have default values but this isn't reflected in the produced signatures :/

Since `inspect.getargs()` is [used a fair bit in third-party projects](https://grep.app/search?q=inspect.getargs&words=true&filter[lang][0]=Python). In the vast majority of these uses, they already have access to the function object, so they should probably just use `Signature.from_callable()` instead of this new method. But there are a few uses where they don't have access to the function object -- so I agree that something like this new method _is_ probably needed, as a replacement for those uses:
- https://github.com/0compute/xtraceback/blob/5f4ae11cf21e6eea830d79aed66d3cd91bd013cd/xtraceback/xtracebackframe.py#L53-L55
- https://github.com/ponyorm/pony/blob/6f7620c9e66a4b8d4f7f7ba4a456b733aab9f374/pony/orm/decompiling.py#L793-L798
- https://github.com/parrt/lolviz/blob/301779d39934937795b71ad5d9d62d8a45f7bee1/lolviz.py#L297-L303

Reply:
> the signatures produced for `func1`, `func2` and `func5` here are all incorrect

Hmmm... well, they are actually valid signatures _for the code object_, since if you pass the code object to `exec()`, it doesn't take any notice of whether the original function had a default value or not:

```pycon
>>> def foo(x=42): print(x)
...
>>> foo()
42
>>> c = foo.__code__
>>> exec(c)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() missing 1 required positional argument: 'x'
```

Why anybody would care what the signature of the code object itself is, though, I'm not sure -- ISTM that everybody trying to get the signature from a code object is just using the code object as the next-best-thing for the function the code object belongs to. I don't think many people are passing code objects directly to `exec()`.

Reply:
Maybe we just need to flag more forcefully in the docs the limitations of this constructor method. I can offer some suggestions once you've addressed the review comments I've already posted 👍

Reply:
I added a `.. note::` notation to highlight it even better.

Reply:
The more I think about this, the more concerned I am :/

The existing `getargs()` function that we have in `inspect` gives _accurate_ information. It doesn't _claim_ to be able to tell you everything about the signature: it only _claims_ to be able to tell you _some_ information about the signature:
- It gives you a list of argument names
- It tells you the name of the `*vararg` argument, if there is one
- It tells you the name of the `**kwarg` argument, if there is one

This new function, however, _claims_ to be able to give you a complete and accurate signature from a code object. That's impossible, since code objects don't have information about which arguments have default values, and so the function gives you _inaccurate information_.

I'm not sure having a prominent note in the docs is sufficient to fix this issue: adding a new constructor that we know gives inaccurate results seems very troubling to me. Perhaps a better solution would be to document `getargs` and promote it to public API?

Reply:
`getargs` is fundamentally broken:

```python
>>> import inspect
>>> def a(x, /, *, y): ...

>>> inspect.getargs(a.__code__)
Arguments(args=['x', 'y'], varargs=None, varkw=None)
```

It does not give you *any* helpful information about `x` and `y`.
So, promoting it to a signature is much better:

```python
>>> inspect.Signature.from_code(a.__code__)
<Signature (x, /, *, y)>
```

You can actually know the parameters' kinds. Now, you can actually call it:

```python
>>> import types
>>> types.FunctionType(a.__code__, {}, 'a')(1, y=2)
```

Compare it with the failed attempt, when you *don't* know the parameter kinds:

```python
>>> types.FunctionType(a.__code__, {}, 'a')(x=1, y=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: a() got some positional-only arguments passed as keyword arguments: 'x'
```

Reply:
True. This new API would provide helpful information that is not currently offered by `getargs`.

However. Does it make sense to think of code objects even having a "signature"? _Functions_ have signatures, and you can _construct_ functions from code objects. But as your snippet above demonstrates, you can't directly call code objects (except via `exec()`, but nobody does that really) -- you have to create a new function from the code object in order to get something that can be called.

Which "signature" are we trying to get here? The signature of the function from which the code object originally came (if it came from a code object), or the signature a new function object would have if we created one from the code object? Getting a "signature from a code object" feels like quite a confusing concept to me.

Reply:
Another option: we could document that if people want to get "signature information" from a code object `co`, they just need to do this:

```py
inspect.signature(types.FunctionType(co, {}))
```

That way, they are forced to be explicit that code objects don't really have a "signature": you need to construct a new function from the code object in order to get a code object that actually has a signature

Reply:
```python
>>> import types, inspect
>>> def a(x: int, /, *, y: str = '') -> None: ...
... 
>>> inspect.signature(types.FunctionType(a.__code__, {}))
<Signature (x, /, *, y)>
```

It is up to you to decide whether or not we should do it. I consider this PR done from my side.

Reply:
Currently, I'm leaning towards not adding this alternative constructor, and documenting that people should call `inspect.signature(types.FunctionType(co, {}))` if they need to get signature information from a code object, and the function object is not available for whatever reason.

One of the third-party projects that uses `getargs` is Quora: https://github.com/quora/qcore/blob/e12a4929c8e75ec2dd878c06270cedcac72a643e/qcore/inspection.py#L130-L148. I'd be curious to know if @JelleZijlstra has any thoughts on whether `getargs` should be documented or deprecated and, if the latter, whether we should add this `Signature.from_code` method or just add documentation along the lines of my sugestion.